lib/chef/provider/package/yum.rb

```diff
@@ -1081,22 +1082,24 @@ def load_current_resource
                 @new_resource.version($2)
               end
             end
-          end
-
-          if @new_resource.version
-            new_resource = "#{@new_resource.package_name}-#{@new_resource.version}#{yum_arch}"
+            @candidate_version << @new_resource.version
+            installed_version << @yum.installed_version(@current_resource.package_name, arch)
```

Comment:
Well that actually doesn't make any more sense in why-run either.

It should just raise if package_name is an array and source is given.  Since that really is a malformed resource it could raise anywhere in the initializer, load_current_resource or define_resource_requirements.  Since raising in the initializer is generally rude, I'd suggest doing it early in load_current_resource.  It doesn't make much sense to try to load the current resource when the resource is malformed like that.

While you're at it, if package_name is an array then the version argument must be either nil or an array of the same length.  Same comments.